### PR TITLE
Added web-push support

### PR DESCRIPTION
**Pull request by ziccardi**
# Motivation
**JIRA**: https://issues.jboss.org/browse/AEROGEAR-9895

### Description

- Added support for the webpush
- Changed the library so that it can used without `cordova` (for webpush)
- Refactored the code adding some OO structure

##### Checklist

- [x] `npm test` passes
- [x] `npm run build` works
- [x] tests are included
- [ ] documentation is changed or added

### CODE REVIEW

**Comment by wtrocki** [comment]
@ziccardi Initial idea of the package was that it will handle registration only and it could be used in cordova and other plugins. Then we kinda brought some Cordova specific code here, but we might revise this completely and move out cordova and have a separate packages for webpush. 

Generally, all depends on how much code/changes/dependencies we will need to add here.
At this point, it looks like the amount of the changes will be small and we are ok with this.

File: packages/push/src/PushRegistrationInterface.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+}
+
+export interface PushRegistrationWebpushOptions extends PushRegistrationOptions {
+  serviceWorker?: string;
+}
+
+/**
+ * Interface for PushRegistration objects.
+ */
+export interface PushRegistrationInterface {
```
Comment (danielpassos) [inline packages/push/src/PushRegistrationInterface.ts]: I know what you did here, but to be very honest for this case I think we don't need a `PushRegistrationInterface`

File: packages/push/src/PushRegistration.ts [hunk 2]
```diff
@@ -24,249 +22,38 @@ export interface PushRegistrationOptions {
...
-          resolve();
-        })
-        .catch(reject);
-      } else {
-        // It should never happend but...
-        return reject(new Error("Push is not properly configured"));
-      }
-    });
+  public async unregister(): Promise<void> {
+    this.delegate.unregister();
```
Comment (danielpassos) [inline packages/push/src/PushRegistration.ts]: Do we need to add an `await` here like in `register()`?

File: packages/push/src/PushRegistration.ts [hunk 3]
```diff
@@ -24,249 +22,38 @@ export interface PushRegistrationOptions {
...
-  public static readonly REGISTRATION_DATA_KEY = "UPS_REGISTRATION_DATA";
+export class PushRegistration implements PushRegistrationInterface {
 
+  /**
+   * Calls to this method are ignored if webpush are used.
+   * @param onMessageReceivedCallback
+   */
   public static onMessageReceived(onMessageReceivedCallback: OnMessageReceivedCallback) {
-    PushRegistration.onMessageReceivedCallback = onMessageReceivedCallback;
+    PushRegistrationCordovaImpl.onMessageReceived(onMessageReceivedCallback);
```
Comment (danielpassos) [inline packages/push/src/PushRegistration.ts]: We should delegate it. Also It's not going to work for WebPush

File: packages/push/src/impl/PushRegistrationWebpushImpl.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+    if (this.validationError) {
+      throw new Error(this.validationError);
+    }
+
+    if (this.httpClient) {
+      const subscription = await this.subscribeToWebPush(this.platformConfig.appServerKey, serviceWorker);
+      const postData = {
+        "deviceToken": JSON.stringify(subscription),
+        "deviceType": "ChromeBrowser",
```
Comment (danielpassos) [inline packages/push/src/impl/PushRegistrationWebpushImpl.ts]: Maybe we should think in a way to get the browser name

File: packages/push/src/impl/PushRegistrationWebpushImpl.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        "operatingSystem": "WebPush",
+        "osVersion": "6.1.2",
```
Comment (danielpassos) [inline packages/push/src/impl/PushRegistrationWebpushImpl.ts]: Maybw we should think in a way to get the browser version

File: packages/push/src/impl/PushRegistrationWebpushImpl.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      await this.httpClient.post(AbstractPushRegistration.API_PATH, postData);
+      const storage = window.localStorage;
+      storage.setItem(AbstractPushRegistration.REGISTRATION_DATA_KEY, JSON.stringify(postData));
+    } else {
+      // It should never happend but...
+      throw new Error("Push is not properly configured");
+    }
+  }
+
+  // tslint:disable-next-line: no-empty
```
Comment (danielpassos) [inline packages/push/src/impl/PushRegistrationWebpushImpl.ts]: Remove?

File: packages/push/src/PushRegistration.ts (hunk 3 above)
Comment (ziccardi) [reply packages/push/src/PushRegistration.ts]: @danielpassos It delegates it to the `PushRegistrationCordovaImpl`. In the comment it is documented that calls to this method are ignored for WebPush.

The whole goal of using delegates here is to make the outer interface unchanged and, unfortunately, the old interface had this static method, so I had to leave it there to keep the interface unchanged.

File: packages/push/src/PushRegistration.ts (hunk 2 above)
Comment (ziccardi) [reply packages/push/src/PushRegistration.ts]: 👍

File: packages/push/src/PushRegistration.ts (hunk 3 above)
Comment (ziccardi) [reply packages/push/src/PushRegistration.ts]: I added a `console.warn` message that is visualised if calling that method while using webpush

File: packages/push/src/impl/PushRegistrationWebpushImpl.ts (hunk 4 above)
Comment (secondsun) [reply packages/push/src/impl/PushRegistrationWebpushImpl.ts]: There's `window.navigator.userAgent` which may be useful.

**Comment by secondsun** [comment]
Just a quick note, 

Chrome and firefox both work great on Windows with this.

Edge can register just fine, but unregistration causes a problem due to the "?" token in the endpoint field in the registration info.  We use this registration info in place of the device token.

Perhaps we should discuss transforming the webpush registration information into some way (base64 encode?) and change how the server treats the it.

File: packages/push/src/PushRegistration.ts (hunk 3 above)
Comment (ziccardi) [reply packages/push/src/PushRegistration.ts]: No need to have that comment anymore: with the latest changes it is now used for webpush too.

**Comment by ziccardi** [comment]
> Just a quick note,
> 
> Chrome and firefox both work great on Windows with this.
> 
> Edge can register just fine, but unregistration causes a problem due to the "?" token in the endpoint field in the registration info. We use this registration info in place of the device token.
> 
> Perhaps we should discuss transforming the webpush registration information into some way (base64 encode?) and change how the server treats the it.

Done!

**Pull request merged**